feat: Adds support for SPL transaction command

Macbeth98 opened the pull request:
# Description
- Adds Support for SPL Transaction Command.
- Parsing of Transaction command with optional Arguments such as `startswith`, `endswith`, and list of fields is done.
- The parsed arguments are available in the `QueryAggregator` struct as `TransactionArguments`.
- Need to work on the actual implementation of the command.
#### Updates
- Implemented the preliminary version of the SPL Transaction command.

Fixes #281 

# Testing
- Written the test cases for the parsing of the grammar for the transaction command in `splParser_Test.go`
#### Updates
- Tested with GUI by running Transaction commands.
- Tested by writing test cases for the implementation of SPL Transaction in `searchHandler_test.go`


# Checklist:

- [X] I have self-reviewed this PR.
- [X] I have removed all print-debugging and commented-out code that should not be merged.
- [X] I have added sufficient comments in my code, particularly in hard-to-understand areas.
- [X] I have formatted the code, if applicable. For Go, I have run `goimports -w .`.

AndrewHess commented on pkg/ast/spl/spl.peg:
```diff
@@ -2030,6 +2036,107 @@ Subseconds <- ("us" / "ms" / "cs" / "ds") {
+
+// Parses one or more FieldNames separate by a space
+// Returns a slice containing all the FieldNames
+FieldList <- first:FieldName rest:(SPACE FieldName)* {
```
There's already a `FieldNameList` that is comma separated. Since this has to be space separated, maybe rename this new rule something like `SpaceSeparatedFieldNameList`

Macbeth98 replied:
I have the name as you have suggested.

AndrewHess commented on pkg/ast/spl/spl.peg:
```diff
@@ -2030,6 +2036,107 @@ Subseconds <- ("us" / "ms" / "cs" / "ds") {
+
+    restSlice := rest.([]any)
+    txnArgs := &TxnArgs{
+        argOption: "txn-definition",
```
Is `"txn-definition"` just used as an invalid default? If so it would be good to rename it to something like `txn-invalid-option` so it's clearer that it is invalid. Or maybe I'm missing something?

Macbeth98 replied:
- The arguments `startswith` and `endswith` are classified as `txn-definition` options in Splunk. And there are more like `maxpause`, `maxspan`, etc. coming under `txn-definition` options. 

- Splunk also supports `memcontrol-options`, `rendering-options`, and `transaction-name` options with its own set of multiple optional arguments. 

- So I just made this classification for now. If we add more arguments depending on the classification on Splunk, we can also similarly group while parsing the spl command.

AndrewHess reviewed (commented):
Looks pretty good so far

AndrewHess commented on pkg/ast/spl/tests/splParser_test.go:
Nice testing

AndrewHess commented on pkg/ast/spl/spl.peg:
```diff
@@ -2047,39 +2048,28 @@ TransactionBlock <- PIPE CMD_TRANSACTION txnOptions:TransactionOptions {
-    if fields != nil {
-        transactionRequest.Fields = fields.([]string)
-    }
+    //if fields != nil {
+      //  transactionRequest.Fields = fields.([]string)
+    //}
```
nit: for the final version of the PR, delete these lines if they're not needed

Macbeth98 replied:
Yes, I will delete them!

AndrewHess reviewed (commented):
The parsing looks good

Macbeth98 commented:
@AndrewHess @nkunal 
- Implemented the preliminary version of the SPL Transaction command in `searchHandler.go`
- Tested with GUI by running Transaction commands.
- Tested by writing test cases for the implementation of SPL Transaction in `searchHandler_test.go`

- Let me know of your thoughts, changes, or any suggestions. Thank you!

nkunal commented on pkg/ast/pipesearch/searchHandler.go:
```diff
@@ -459,6 +459,10 @@ func convertRRCsToJSONResponse(rrcs []*sutils.RecordResultContainer, sizeLimit u
 	}
 
 	allJsons, allCols, err := record.GetJsonFromAllRrc(rrcs, false, qid, segencmap, aggs)
+	if aggs != nil && aggs.TransactionArguments != nil && err == nil {
```
the below err != nil check shoule be before this if block: a practice we follow, always process the err first and then continue with the rest of the processing

Macbeth98 replied:
Yes got it! I will change it!

nkunal reviewed (commented):
nit: remove the unneeded white spaces that got added in the PR.

AndrewHess commented on pkg/ast/spl/spl.peg:
```diff
@@ -2030,6 +2037,112 @@ Subseconds <- ("us" / "ms" / "cs" / "ds") {
+    return txnArg, nil
+}
+
+StartsWithOption <- "startswith" EQUAL str:QuotedString {
+    txnArg := &TxnArgs{
+        argOption: "startswith",
+        startsWith: removeQuotes(str.(string)),
+    }
+    return txnArg, nil
+}
+
+EndsWithOption <- "endswith" EQUAL str:QuotedString {
```
The Splunk documentation says the startswith and endswith values are "a search or eval filtering expression". I think for the initial PR you can keep these as they are, but eventually we'll need to support things like `startswith=(apple OR "banana")`. We should be able to reuse some of the existing rules in this file to handle that parsing.

AndrewHess reviewed (commented):
It's looking good so far. I think you'll have to do the transaction logic in segaggs.go though so that queries can chain the output into another command

Macbeth98 commented on pkg/ast/pipesearch/searchHandler.go:
```diff
@@ -459,6 +459,10 @@ func convertRRCsToJSONResponse(rrcs []*sutils.RecordResultContainer, sizeLimit u
+	if aggs != nil && aggs.TransactionArguments != nil && err == nil {
+		// Process Splunk Transaction command.
+		allJsons, allCols, err = processTransactionsOnRecords(allJsons, allCols, aggs.TransactionArguments)
+	}
```
Yes got it, I will change the location of the processing!

Macbeth98 commented on pkg/ast/spl/spl.peg:
```diff
@@ -2030,6 +2037,112 @@ Subseconds <- ("us" / "ms" / "cs" / "ds") {
+    return queryAgg, nil
+}
+
+// The List of fields should Always be given at the last.
```
Yes, I will remove it!

AndrewHess commented on pkg/segment/reader/record/rrcreader.go:
```diff
@@ -181,6 +181,24 @@ func GetJsonFromAllRrc(allrrc []*utils.RecordResultContainer, esResponse bool, q
 	sort.Strings(colsSlice)
 	log.Infof("qid=%d, GetJsonFromAllRrc: Got %v raw records from files in %+v", qid, len(allRecords), time.Since(sTime))
+
+	if aggs != nil {
+		for post := aggs; post != nil; post = post.Next {
+			if post.TransactionArguments != nil {
+				recs := make(map[string]map[string]interface{}, 0)
+				recs["0"] = make(map[string]interface{})
+				recs["0"]["records"] = allRecords
+				recs["0"]["columns"] = colsSlice
+				nodeRes := &structs.NodeResult{}
+				agg.PostQueryBucketCleaning(nodeRes, aggs, recs, finalCols)
+				allRecords = recs["0"]["records"].([]map[string]interface{})
+				colsSlice = recs["0"]["columns"].([]string)
+				break
+			}
+		}
+
+	}
```
There's some similar logic where we call `PostQueryBucketCleaning()` above this on line 112. If you're able to reuse that logic I think that would be better so the code doesn't get too bloated if we add more commands that need to do something like this.

AndrewHess reviewed (commented):
This is looking good. Just one thing to fix that I see

AndrewHess commented on pkg/ast/spl/spl.peg:
```diff
@@ -2048,8 +2053,210 @@ Subseconds <- ("us" / "ms" / "cs" / "ds") {
+    return txnArg, nil
+}
+
+// Rerturns structs.FilterStringExpr
```
nit: typo for "Returns"

Macbeth98 commented:
@AndrewHess @nkunal Please review the PR whenever you can. Thank You!

nkunal commented:
we should add some tests in the cicd/ingest.csv to have an end to end test for this spl command

AndrewHess commented on pkg/segment/aggregations/segaggs_test.go:
```diff
@@ -0,0 +1,605 @@
+}
+
+// Test Cases for processTransactionsOnRecords
+func All_TestCases() (map[int]bool, []*structs.TransactionArguments) {
+	recordsLengthPositive := make(map[int]bool)
```
Should this function be named something specific to transactions, or is intended for other uses as well later? Also what does the name `recordsLengthPositive` mean?

Macbeth98 replied:
I am checking that to see whether this particular will give results `greater than 0(positive) or nt`. For each test case command, I added this to check that.

So I named it `recordsLengthPositive` indicating that. I will add comments beside that. Also, should I change the name to something else like `isRecordsLengthPositive` or `isRecordsCountGtZero`?

AndrewHess replied:
Ok make sense. Something like `matchesSomeRecords` feels more descriptive to me, but I think it's just a personal preference thing, so you can do what makes sense to you.

Should the `All_TestCases` be renamed `All_TransactionTestCases`, or will later test cases for other commands potentially add their test cases here as well?

Macbeth98 replied:
Yes, I named it to `All_TestCasesForTransactionCommands`, as I want to keep this for Transaction Test cases only.
Yes, I think `matchesSomeRecords` makes more sense to me, also. I will change the variable name to that.

AndrewHess commented on pkg/segment/aggregations/segaggs_test.go:
```diff
@@ -0,0 +1,605 @@
+				if txnArgs.StartsWith != nil {
+					if ind == 0 {
+						if txnArgs.StartsWith.StringValue != "" {
+							assert.Equal(t, eventMap["http_method"], txnArgs.StartsWith.StringValue)
+						} else if txnArgs.StartsWith.StringClauses != nil {
+							assert.Contains(t, txnArgs.StartsWith.StringClauses[0][0], eventMap["http_method"])
```
You're testing on `http_method` just because `http_method` values are what you happened to use for the txnArgsN variables, right?

Macbeth98 replied:
Yes, Every, test case I am giving `startswith` command to be related to `http_method`. So that it would make the testing of different variations of command easier.

But, I realized I need to change my test case to handle other cases. I am handling two cases, but I need to add the case for eval and search. I will add them.

Macbeth98 replied:
I have added more test cases and also the functions that verify the results.

AndrewHess reviewed (commented):
Looks good overall, nice work on this and on adding thorough unit tests

Macbeth98 commented on pkg/ast/spl/spl.peg:
```diff
@@ -2048,8 +2091,198 @@ Subseconds <- ("us" / "ms" / "cs" / "ds") {
+}
+
+// eval(<bool-expression>)
+TransactionEval <- "eval" EMPTY_OR_SPACE L_PAREN condition:BoolExpr R_PAREN {
```
I am not sure, I thought users would write `eval ()`. But if you want me to remove the space, I will remove it.

Macbeth98 replied:
Hey @AndrewHess , I just checked by executing on Splunk cloud. Space can be there after the eval keyword like this `eval ()`.

Macbeth98 commented on pkg/ast/spl/spl.peg:
```diff
@@ -131,6 +131,44 @@ func getRexColNames(pattern string) ([]string, error) {
+		exprType = utils.SS_DT_STRING
+	case int, int32, int64, float32, float64, json.Number:
+		value = json.Number(reflect.ValueOf(v).String())
+		exprType = utils.SS_DT_SIGNED_NUM
```
I did that to `SS_DT_SIGNED_NUM`, to group all number types and in the process of the command, I am converting those number types to Float Types. I guess, I can change this to `SS_DT_FLOAT` here only.
But I did this way as I needed to convert the values from the record to float.

Macbeth98 replied:
Just to make are you asking me to assign the exprType to`SS_DT_FLOAT` or you want me to convert the value to float instead of doing `json.Number()`.

Macbeth98 commented on pkg/ast/spl/spl.peg:
```diff
@@ -2048,8 +2091,198 @@ Subseconds <- ("us" / "ms" / "cs" / "ds") {
+}
+
+// Returns structs.FilterStringExpr
+TransactionFilterString <- strExpr:(TransactionQuotedString / TransactionEval / TransactionSearch / TransactionQuotedStringClausesExpr) {
```
This is for the case where `eval` keyword is not used. So `startswith=("st1" or "st2" and "st3")`; something like this.

AndrewHess replied:
Whoops I meant `TransactionSearch`, not `TransactionEval`. But I see now that `TransactionSearch` just handles one search term, not a group of them with `AND` and `OR`. My main point on this is that is seems like `TransactionQuotedStringClausesExpr` duplicates a lot of parsing logic, so I was wondering if just changing
```
TransactionSearch <- expr:SearchTerm {
    // Make the simpleSearchExpr from the ast.Node
}
```
to
```
TransactionSearch <- expr:ClauseLevel1 {
    // Make the simpleSearchExpr from the ast.Node
}
```
would remove the need for the `TransactionQuotedStringClausesExpr` rule. If so I think removing that (and the rules that only this rule uses) would clean it up a bit. However, if that won't work, what you have is fine

Macbeth98 replied:
Yes, I am planning to make this fix. I will try to fix this.

Macbeth98 replied:
@AndrewHess This is properly cleaned up now. I have removed `TransactionQuotedStringClausesExpr` and also I converted the `TransactionSearch <- expr:ClauseLevel1` and I kept that `ast.Node` and converted to `ASTNode` later like how I am doing for `TransactionQuotedStringSearchExpr`. 

With this, I have removed the options of `StringClauses` and `SearchTerm` which will make the `SimpleSearchExpr` is not required.

Macbeth98 commented on pkg/ast/spl/spl.peg:
```diff
@@ -2048,8 +2091,198 @@ Subseconds <- ("us" / "ms" / "cs" / "ds") {
+    return strs, nil
+}
+
+TransactionSearch <- expr:SearchTerm {
+    astNode := expr.(*ast.Node)
+
+	simpleSearchExpr, err := resolveTerminalAstNodeToSimpleSearch(astNode)
+	if err != nil {
```
Yes, I will fix that.

Macbeth98 commented:
@AndrewHess 

- I have removed and cleaned up structs and code surrounding `TransactionQuotedStringSearchExpr`, `TransactionSearch`, and `SimpleSearchExpr`. 

- I have added more test cases on the parser side and also on the implementation side I have added functions that verify the results of the transaction command output.

- I believe, the PR is ready to be merged.

Let me know of any changes. Thank you!

AndrewHess reviewed (approved):
Looks good
